src/crimson/common/gated.h

```diff
@@ -0,0 +1,61 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#pragma once
+
+#include <seastar/core/gate.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/future-util.hh>
+
+#include "crimson/common/exception.h"
+#include "crimson/common/log.h"
+#include "include/ceph_assert.h"
+
+namespace crimson::common {
+
+class Gated {
```

Comment:
Actually, since we want each component to handle their own ongoing event processing, each of them needs a gate to guard those processings, so I put thesse common operations in this class. It is not supposed to be related to system_shutdown_exception.

Reply:
> * `print()` is originally used by `Messenger` internally for a special purpose to reuse Messenger prefix inside Connection prefix. I'm not sure if it is also reasonable to be used here. And seems no one uses it correctly.
> 
I added this print() method to print the exact component that triggered an exception.

Reply:
OK, then I think these needs to be addressed:
* `print()` should be implemented correctly for sub-classes of `Gated`;
* It would be duplicated if components have multiple gates;